File: src/c99/parser.c
/*********************************************************************************************************************
 *					   Space Science & Engineering Lab - MSU
 *					     Maia University Nanosat Program
 *
 *							IMPLEMENTATION
 * Filename	  : link.c
 * Created	  : 15, Nov 2006
 * Description:
 **********************************************************************************************************************/
#include "parser.h"

int initParser(Parser* parser, void (*packet_callback)(Packet*))
{
  parser->state = STATE_FINDSTART;
  parser->delimCount = 0;
  parser->controlFlags = 0;
  parser->headerIndex = 0;
  parser->headerLength = 0;
  parser->dataIndex = 0;
  parser->dataLength = 0;
  parser->crcIndex = 0;
  parser->crcLength = 0;
  parser->packet_callback = packet_callback;
  return 0;
}

int acceptPacket(Parser* parser)
{
  Packet packet; // declare memory on the stack
  initPacket(&packet);

  // populate a Packet struct from the raw data
  readPacketFromBuffer(&packet, parser->packetBuffer);

  // deliver to application code
  if (parser->packet_callback) {
    parser->packet_callback(&packet);
  } else {
    return -3; // TODO enumerate errors
  }
  parser->state = STATE_FINDSTART;
  parser->headerIndex = 0;
  parser->delimCount = 0;

  return 0;
}

// consumes a byte stream and generates calls to packet_callback
int parseBytes(Parser* parser, INT08U* buffer, int numBytes)
{
  INT08U msg;
  // consume input one byte at a time
  for (int i = 0; i < numBytes; i++)
  {
    msg = buffer[i]; // read next byte in sequence

    // check for escape char (occurs mid-frame)
    if((parser->delimCount == (FRAME_LEADER_LENGTH-1)) && (msg == FRAME_ESCAPE))
    { // reset FRAME_LEADER detection
      parser->delimCount = 0;
      continue;
    }
    else if(msg == FRAME_LEADER)
    {
      parser->delimCount++;
      if(parser->delimCount == FRAME_LEADER_LENGTH)
      {
        for (int i = 0; i < FRAME_LEADER_LENGTH; i++)
          parser->packetBuffer[i] = FRAME_LEADER;
        parser->headerIndex = FRAME_LEADER_LENGTH;
        parser->state = STATE_GET_CF;
        continue;
      }
    }
    else /* reset delim count */
      parser->delimCount = 0;

    /* use current char in following state */
    switch(parser->state)
    {
    case STATE_FINDSTART:
        /* Do Nothing, dump char */
      break; /* end STATE_FINDSTART */

    case STATE_GET_CF:
      if(parser->headerIndex>=MAX_HEADER_SIZE)
      {   // TODO error reporting
        parser->state = STATE_FINDSTART;
        break;
      }
      parser->packetBuffer[parser->headerIndex] = msg;
      parser->headerIndex++;
      if(parser->headerIndex==6)
      { /* Ham CF */
        INT16U cf = readINT16U(&parser->packetBuffer[FRAME_LEADER_LENGTH]);
        parser->controlFlags = CFHamDecode(cf);
        writeINT16U(&parser->packetBuffer[FRAME_LEADER_LENGTH], parser->controlFlags);
        parser->headerLength = headerLength(parser->controlFlags);
        parser->state = STATE_GETHEADER;
      }
      break; /* end STATE_GET_CF */

    case STATE_GETHEADER:
      if(parser->headerIndex>=MAX_HEADER_SIZE)
      { // TODO error reproting
        parser->state = STATE_FINDSTART;
        break;
      }
      parser->packetBuffer[parser->headerIndex] = msg;
      parser->headerIndex++;
      if(parser->headerIndex>=parser->headerLength)
      {
        if (parser->controlFlags & CF_DATA)
        {
          parser->dataIndex = 0;
          int dataOffset = headerFieldOffset(parser->controlFlags, CF_DATA);
          parser->dataLength = readINT16U(&parser->packetBuffer[dataOffset]);
          parser->data = parser->packetBuffer + parser->headerLength; // set the data pointer to an offset in packetBuffer
          parser->state = STATE_GETDATA;
        }
        else if (parser->controlFlags & CF_CRC)
        {
          parser->crcIndex = 0;
          parser->dataLength = 0;
          parser->crcLength = CRC_FIELD_SIZE;
          parser->crcSum = parser->data + parser->dataLength;
          parser->state = STATE_GETCRC;
        }
        else
        {
          acceptPacket(parser);
        }
      }
      break; /* end STATE_GETHEADER */

    case STATE_GETDATA:
      parser->data[parser->dataIndex] = msg;
      parser->dataIndex++;

      if(parser->dataIndex >= parser->dataLength)
      {
        if (parser->controlFlags & CF_CRC)
        {
          parser->crcIndex = 0;
          parser->crcLength = CRC_FIELD_SIZE;
          parser->crcSum = parser->data + parser->dataLength;
          parser->state = STATE_GETCRC;
        }
        else {
          acceptPacket(parser);
        }
      }
      break; /* end STATE_GETDATA */

    case STATE_GETCRC:
      parser->crcSum[parser->crcIndex] = msg;
      parser->crcIndex++;
      if(parser->crcIndex >= parser->crcLength)
      {
        int checkSumInputLength = parser->headerLength+parser->dataLength;
        INT32U expectedCRC = readINT32U(parser->crcSum);
        INT32U computedCRC = getCRC(parser->packetBuffer, checkSumInputLength);
        if ( expectedCRC != computedCRC)
        { // TODO error reporting
          parser->state = STATE_FINDSTART;
        } else {
          acceptPacket(parser);
        }
      }
      break; /* end STATE_GETCRC */
    } /* end switch(state) */
  }/* end while(forever) */
  return numBytes;
}

File: src/csharp/ELPLib/Parser.cs
/**********************************************************************************************************************
 *                                         Space Science & Engineering Lab - MSU
 *                                           Maia University Nanosat Program
 *
 *                                                    IMPLEMENTATION
 * Filename      : Link.cs
 * Programmer(s) : Grant Nelson, Chad Bohannan
 * Created       : 22 Aug, 2006
 * Description   : Manages a connection over a channel using SSEL Packets.
 **********************************************************************************************************************/
using System.Collections;
using System.ComponentModel;
using System.Threading;
using System;

namespace ELP
{
    public interface IPacketHandler
    {
        void OnPacket(Packet packet);
    }


    /// <summary>
    /// Parser class buffers content and maintains the state machine variables 
    /// </summary>
    public class Parser
    {
        #region Types
        // These are the states that the parser can be in.
        private enum State
        {
            FindStart,
            GetCF,
            GetHeader,
            GetData,
            GetCRC
        }
        
        #endregion

        #region Attrubutes

        private byte delimCount; // The current count of sequental delimiters.
        private State state; 
        byte[] buffer; // byte buffer a Packet is parsed from
        private ushort controlFlags;
        private int headerIndex;
        private int headerLength;
        private int dataIndex;
        private int dataLength;
        private int crcIndex;
        const int crcLength = 4;
        const int MaxHeaderSize = 19;
        const int MaxDataSize = 1 << 16;        
        const int MaxPacketSize = MaxHeaderSize + MaxDataSize + crcLength;

        IPacketHandler handler;

        #endregion

        public Parser(IPacketHandler handler)
        {
            this.handler = handler;
            buffer = new byte[MaxPacketSize];
            reset();
        }

        public void reset()
        {
            state = State.FindStart;
            headerIndex = 0;
            headerLength = 0;
            dataIndex = 0;
            dataLength = 0;
            crcIndex = 0;
        }

        public int readBytes(byte[] data)
        {
            foreach (byte b in data)
            {
                // a simple state machine to strip framming bytes from incoming data
                switch(b)
                {
                    case Packet.FrameLeader:
                        delimCount += 1;
                        if (delimCount >= Packet.FrameLeaderLength) // frame leader detected
                        {
                            reset();
                            continue;
                        }
                        break;
                    case Packet.FrameEscape:
                        if (delimCount == Packet.FrameLeaderLength - 1)
                        {
                            delimCount = 0;
                            continue; // drop frame-escape char independent of the packet parser state
                        }
                        goto default; // explicit fallthrough
                    default: // byte is not a FrameLeader
                        if (delimCount >= Packet.FrameLeaderLength) // first packet byte detected
                        {
                            state = State.GetCF;
                            delimCount = 0;
                        }
                        break;
                }

                // a more complex state machine to parse a packet from an input stream
                switch (state)
                {
                    case State.FindStart:
                        break; // drop framing bytes
                    case State.GetCF:
                        GetCF(b); // read
                        break;
                    case State.GetHeader:
                        GetHeader(b);
                        break;
                    case State.GetData:
                        GetData(b);
                        break;
                    case State.GetCRC:
                        GetCRC(b);
                        break;
                }
            }
            return data.Length;
        }

        /// <summary>
        /// This gets the Control Feild and calculates the total header size.
        /// </summary>
        /// <param name="b">The current byte being read.</param>
        private void GetCF(byte b)
        {
            buffer[headerIndex] = b;
            headerIndex += 1;
            if (headerIndex == 2)
            {
                controlFlags = Packet.readUINT16(buffer, 0);
                headerLength = Packet.headerLength(controlFlags);
                if (headerLength > 2)
                {
                    state = State.GetHeader;
                }
                else
                {
                    headerIndex = 0;
                    state = State.FindStart;
                }
            }           
        }

        /// <summary>
        /// This gets the Header and locates the Error Detection Type to find the CRC size.
        /// </summary>
        /// <param name="b">The current byte being worked on.</param>
        private void GetHeader(byte b)
        {
            buffer[headerIndex] = b;
            headerIndex += 1;
            if (headerIndex == headerLength)
            {
                if ((controlFlags & (ushort)Packet.ControlFlag.data) != 0)
                {
                    dataLength = Packet.readUINT16(buffer, headerIndex - 2);
                    dataIndex = 0;
                    state = State.GetData;
                }
                else if ((controlFlags & (ushort)Packet.ControlFlag.crc) != 0)
                {
                    dataLength = 0;
                    dataIndex = 0;
                    crcIndex = 0;
                    state = State.GetCRC;
                }
                else
                {
                    handler.OnPacket(new ELP.Packet(buffer));
                    reset();
                }
            }
        }

        /// <summary>
        /// This gets the body's data then sets the CRC size for the body.
        /// </summary>
        /// <param name="b">The current byte being worked on.</param>
        private void GetData(byte b)
        {
            int offset = headerIndex + dataIndex;
            buffer[offset] = b;
            dataIndex += 1;
            if (dataIndex == dataLength)
            {
                if ((controlFlags & (ushort)Packet.ControlFlag.crc) != 0)
                {
                    crcIndex = 0;
                    state = State.GetCRC;
                }
                else
                {
                    handler.OnPacket(new ELP.Packet(buffer));
                    reset();
                }
            }
        }

        /// <summary>
        /// This gets the CRC, checks the body then sends the package before reseting the parser.
        /// </summary>
        /// <param name="b">The current byte being worked on.</param>
        private void GetCRC(byte b)
        {
            int offset = headerIndex + dataIndex + crcIndex;
            buffer[offset] = b;
            crcIndex += 1;
            if (crcIndex == crcLength)
            {
                // TODO inspect CRC value

                handler.OnPacket(new ELP.Packet(buffer));
                reset();
            }
        }

    }// Parser class
}
